osbuild/util/containers.py

```diff
@@ -127,4 +148,32 @@ def container_source(image):
             os.symlink(image_filepath, tmp_source)
 
         image_source = f"{container_format}:{tmp_source}"
-        yield image_name, image_source
+        yield image_source
+
+
+@contextmanager
+def container_source(image):
+    image_filepath = image["filepath"]
+    container_format = image["data"]["format"]
+    image_name = image["data"]["name"]
+
+    if container_format not in ("dir", "oci-archive", "containers-storage"):
+        raise RuntimeError(f"Unknown container format {container_format}")
+
+    if container_format == "containers-storage":
```

Comment:
(super nitpick^2) the two calls of *_container_storage_source() look very similar, so we could just assign the right function to a variable and only have a single call. this would make the amount of change smaller if we ever need to change the function signatures. But totally optional and I realize the wins are small, I still wanted to mention it :)
(for illustration) something like:
```python
diff --git a/osbuild/util/containers.py b/osbuild/util/containers.py
index da7ce081..4c9ebb63 100644
--- a/osbuild/util/containers.py
+++ b/osbuild/util/containers.py
@@ -157,15 +157,15 @@ def container_source(image):
     container_format = image["data"]["format"]
     image_name = image["data"]["name"]
 
-    if container_format not in ("dir", "oci-archive", "containers-storage"):
-        raise RuntimeError(f"Unknown container format {container_format}")
-
     if container_format == "containers-storage":
-        with containers_storage_source(image, image_name, image_filepath, container_format) as image_source:
-            yield image_name, image_source
+        cnt_src_fn = containers_storage_source
+    elif container_format in ("dir", "oci-archive"):
+        cnt_src_fn = dir_oci_archive_source
     else:
-        with dir_oci_archive_source(image, image_name, image_filepath, container_format) as image_source:
-            yield image_name, image_source
+        raise RuntimeError(f"Unknown container format {container_format}")
+
+    with cnt_src_fn(image, image_name, image_filepath, container_format) as image_source:
+       yield image_name, image_source
 
 
 def get_host_storage(storage_conf):
```
or alternatively
```python
diff --git a/osbuild/util/containers.py b/osbuild/util/containers.py
index da7ce081..13b0697c 100644
--- a/osbuild/util/containers.py
+++ b/osbuild/util/containers.py
@@ -157,15 +157,17 @@ def container_source(image):
     container_format = image["data"]["format"]
     image_name = image["data"]["name"]
 
-    if container_format not in ("dir", "oci-archive", "containers-storage"):
-        raise RuntimeError(f"Unknown container format {container_format}")
-
-    if container_format == "containers-storage":
-        with containers_storage_source(image, image_name, image_filepath, container_format) as image_source:
-            yield image_name, image_source
-    else:
-        with dir_oci_archive_source(image, image_name, image_filepath, container_format) as image_source:
-            yield image_name, image_source
+    dispatcher = {
+        "containers-storage": containers_storage_source,
+        "dir": dir_oci_archive_source,
+        "oci-archive": dir_oci_archive_source,
+    }
+    cnt_src_fn = dispatcher.get(container_format)
+    if not cnt_src_fn:
+       raise RuntimeError(f"Unknown container format {container_format}") 
+
+   with cnt_src_fn(image, image_name, image_filepath, container_format) as image_source:
+       yield image_name, image_source
 
 
 def get_host_storage(storage_conf):
```

Reply:
It went through my mind but didn't love my implementation 😅 I really like the second option though so I'll update that